%************************************************
\chapter{Inference of pseudotimes with switch-like models}\label{ch:ouijachap} % $\mathbb{ZNR}$
%************************************************

\section{Introduction}

A predominant feature of current pseudotime algorithms is that they  emphasise an ``unsupervised''
 approach where pseudotimes are learned using no specific prior knowledge of gene behaviour. Typically, the high-dimensional molecular profiles for each cell are projected on to a reduced dimensional space by using a (non)linear transformation provided by the manifold learning algorithms discussed in section \ref{sec:int:manifold}. In this reduced dimensional space, it is hoped that any temporal variation is sufficiently strong to cause the cells to align against a trajectory along which pseudotime can be measured. This approach is therefore subject to a number of analysis choices, e.g. the choice of dimensionality reduction technique, the trajectory fitting algorithm, etc., that could lead to considerable variation in the pseudotime estimates obtained. In order to verify that any specific set of pseudotime estimates are biologically plausible, it is typical for investigators to retrospectively examine specific marker genes or proteins to confirm that the predicted (pseudo)temporal behaviour reflects \emph{a priori} beliefs. An iterative ``semi-supervised" process maybe therefore be required to concentrate pseudotime algorithms on behaviours that are both consistent with the measured data and compliant with a limited amount of known gene behaviour.

In this chapter we present an orthogonal approach implemented in a latent variable model statistical framework called Ouija that can integrate prior expectations of gene behaviour along trajectories using Bayesian nonlinear factor analysis. Our approach uses known or putative marker genes directly for pseudotime estimation rather than as a device for retrospectively validating pseudotime estimates. In particular, our model focuses on switch-like expression behaviour and assumes that the marker gene expression follows a noisy switch pattern that corresponds to up- or down-regulation over time. Crucially, we explicitly model when a gene turns on or off as well as how quickly this behaviour occurs. We can then place Bayesian priors on this behaviour that allows us to learn temporal gene behaviours that are consistent with existing biological knowledge.

This chapter proceeds by explicitly framing single-cell pseudotimes as a statistical latent variable model, departing from typical approaches that employ a two stage dimensionality reduction and cell ordering process. We consider some desirable properties of such models including their mean function and mean-variance relationship. We benchmark this method on synthetic data before applying it to a range of single-cell developmental trajectories.


\section{Connection between latent variable modelling and trajectories} \label{sec:latvar}

The aim of pseudotime ordering is to associate a $G$-dimensional expression measurement $\mby_n$ of cell $n \in 1, \ldots, N$ to a latent unobserved univariate pseudotime $t_n$. Mathematically we can express this as
\begin{align}
 \underbrace{\mby_n}_{\mathclap{\text{Expression}}} = \underbrace{\mbf_\Theta}_{\text{Mapping}}(\underbrace{t_n}_{\text{Pseudotime}}) + \underbrace{\epsilon_n}_{\text{Noise}}
 \label{eq:master_eq}
\end{align}
where the function $\mbf_\Theta$ maps the one-dimensional pseudotime $t_n$ to the $G$-dimensional observation space in which the data lies with (possibly output-dimension dependent) parameters $\Theta$. The challenge lies in the fact that both the mapping function $f$ and the pseudotimes are \emph{unknown}. Furthermore, it is hard to derive objective criteria to evaluate potential mapping functions - why is $f(t) = \theta t$ (factor analysis) or $f(t) = \text{nonparametric smooth functions}$ (GPLVM) necessarily ``better'' than $f(t) = \text{const}$ or $f(t) = 1 / t$? A key result discussed below is that $f$ corresponds to our \emph{a priori} expectation of how gene expression changes over time.

% Further, the measurement noise $\epsilon_n$ must absorb \emph{all} the variation not explained by the pseudotime and mapping function

\subsection{Factor analysis assumes linear gene activations}

If the mapping functions $\mbf$ are restricted to a linear functions of $t_n$ then the generative model for the expression of gene $g$ in cell $n$ is given by a one-dimensional factor analysis model of the form
\begin{equation} \label{eq:fa}
	\begin{aligned}
		% k_g & \sim \norm(\mu^{(k)}_g, 1 / \tau^{(k)}) \\
		y_{ng} & \sim \norm(\theta_g t_n, \tau^{-1}_g)
	\end{aligned}
\end{equation}

where $\tau_g^{-1}$ is the measurement precision for gene $g$ and the $G-$dimensional vector $\mbtheta$ acts as the factor loading matrix. Note that the data can always be centred making it redundant to model an intercept\footnote{This is only true if the expectation of $t_n$ is 0. We could conceive of a model where we place priors on $t_n$ that correspond to physical capture times, in which case we would need to reintroduce gene-specific intercepts.}.

The interpretation of this model is that the value of $\theta_g$ regulates the response of the gene expression $\mby_{g}$ to the cell's position along the trajectory $t_n$. Given this is necessarily linear, factor analysis assumes a linear change in expression for each gene over pseudotime.

The key insight is that the functional form of $f$ necessarily corresponds to our prior expectations for how genes evolve along pseudotemporal trajectories. For linear $f$, we expect linear behaviour; for nonparametric smooth $f$ (e.g. from GPLVM), we essentially restrict such behaviour to any smooth (mean) function of pseudotemporal progression.

If we modify equation \ref{eq:fa} to model a common precision across all genes so $\tau_g = \tau$ then then this further reduces to
(probabilistic) principal components analysis \cite{tipping1999probabilistic}, providing an explicit interpretation for the results of principal component analysis on single-cell data.

Our objective here is to use parametric forms for the mapping function $f$ that will enable relatively fast computations whilst characterising certain gene expression temporal behaviours. Our premise is that prior knowledge might exist for a set of marker genes whose temporal behaviour is known to be approximately switch-like and therefore can be used to infer pseudo-time orderings.



\section{A generative model of single-cell pseudotime}

\subsection{Modelling considerations}

\begin{figure}% [!t]
	\centering
	\includegraphics[width=1.5\textwidth,angle=90]{gfx/ch3/main_fig}
  \caption{(Caption next page.)} \label{fig:main}
\end{figure}
\addtocounter{figure}{-1}
\begin{figure}
  \caption{{\bf Overview of Ouija.}
  \textbf{A} A small panel of genes is chosen with \emph{a priori} knowledge about their expression dynamics, either as cell-type specific marker genes or from pathway interaction databases such as KEGG. The pseudotimes are subsequently inferred using Ouija followed by downstream analyses such as differential expression and clustering of additional genes  using standard methods such as those described in \cite{Trapnell2014}.
  \textbf{B} Ouija infers pseudotimes using Bayesian nonlinear factor analysis by decomposing the input gene expression matrix through a sigmoidal mapping function. The latent variables become the pseudotimes of the cells while the factor loading matrix is informative of different types of gene behaviour. A heteroskedastic dispersed noise model with dropout is used to accurately model scRNA-seq data.
  \textbf{C}  Examples of sigmoidal gene expression for three different sets of parameters. Points represent simulated gene expression, with the solid black line denoting the sigmoid curve and the red dashed line denoting the activation time.
  The sigmoid function is parameterised by (i) $\mu_0$ - the average peak expression level, (ii) $k$ - the activation strength and (iii) $t_0$ - the activation time. \emph{Left} Fast `switch-on' behaviour with parameters $k= 30$, $\mu_0 = 1$, $t_0 = 0.5$, \emph{Centre} Slower `switch-off' behaviour with parameters $k= -10$, $\mu_0 = 2$, $t_0 = 0.5$ and \emph{Right} Decaying behaviour with parameters $k= -5$, $\mu_0 = 10$, $t_0 = 0$.
  \textbf{D} The effect of prior expectations on the `activation time' $t_0$ parameter. A highly peaked prior (top) corresponds to confident knowledge of where in the trajectory a gene turns on or off, while a more diffuse prior (bottom) indicates more uncertainty as to where in the trajectory the gene behaviour occurs.
  \textbf{E} Comparison of Ouija to alternative pseudotime inference algorithms. Ouija is the only algorithm that explicitly allows incorporation of prior knowledge of gene behaviour.}
\end{figure}


\subsubsection{Mean function}

% When performing dimensionality reduction on scRNA-seq data with the aim of learning a `pseudotime' there is a direct link between the algorithm used and the assumptions of how a gene will behave along the pseudotime trajectory. For example, if a linear dimensionality reduction algorithm is used, such as PCA, ICA or FA, the assumption is that the gene will linearly increase or decrease along the trajectory\footnote{Note that if PCA is used to one dimension then reconstructing the gene expression would result in a rank one matrix. Consequently, the assumption would be that all genes are perfectly correlated or anti-correlated along the trajectory}. Conversely, non-parametric mean functions have also been proposed, such as those used in GP-LVM or smoothing spline models. In the case of GP-LVM, the only restrictions are coded in the form of the covariance matrix, which tend to have assumptions of smoothness or periodicity.

If we wish to build a generative statistical latent variable model of pseudotime we must therefore decide on a sensible set of functions $f$ that accurately characterise gene expression over (pseudo-)time. We have so far essentially considered $f$ as a linear change over time (PCA/factor analysis) and $f$ simply as smooth functions (GPLVM).

However, both the linearly-restricted case and the entirely nonparametric cases are unrealistic for different reasons. The linear case assumes that the mean function must change across the entire trajectory (ie there can be no parts where the mean value is stationary), making modelling of complex processes where different genes are responsible for different parts of the trajectory impossible. Furthermore, it assumes that as the trajectory progresses the expression values will tend to $\pm \infty$ (which has the additional issue of negative expression values). While the linear case is too restrictive, the non-parametric case is overly-flexible, allowing biologically unrealistic expression patterns such as cubic polynomials or genes with multiple modes over a relatively short trajectory.

Instead we opt for a `goldilocks' mean function, where we assume genes are off, undergo a period of near-linear increase and settle at a constant value (or conversely begin on then turn off)\footnote{In practice this restricts us to the class of monotonically increasing and decreasing gene expression signatures}. A good approximation to this is the sigmoid function which parametrises the mean as

\begin{equation}
f(t; \mu_0, k, t_0) = \frac{2 \mu_0}{1 + \exp\left( -k (t - t_0) \right)}
\end{equation}

where $t$ corresponds to the pseudotime of a cell, $\mu_0$ corresponds to the mean expression of the gene, $k$ is the \emph{activation strength} and corresponds to how fast the  gene turns on or off and $t_0$ corresponds to the \emph{activation time} - where along the trajectory the gene behaviour occurs. The key result here is that the activation times and strengths are highly interpretable quantities for which realistic Bayesian priors may be available. For example, a researcher may know how quickly a given gene is up or downregulated over pseudotime, providing the sign and magnitude of $k$. Furthermore, they may know whether such regulation occurs early or late in the trajectory, providing prior information for $t_0$. Crucially, by specifying a variance on the prior knowledge the researcher can encode the perceived certainty of the prior information, which leads to differing prior function draws (figure \ref{fig:main}D).

% We use sigmoid mapping functions that can capture ``switch-like'' behaviour over time as shown in Figure \ref{fig:main}C and are parameterised by key quantities for which estimates (and associated uncertainty) might be useful: the activation strength and the activation time. These correspond to measures of how rapidly the gene expression level changes and when the (in)activation of the gene occurs.

The approach we adopt is therefore a form of latent variable model implemented as \emph{non-linear parametric factor analysis} where the factors correspond to the pseudo-times and the factor loadings correspond to the parameters of the sigmoidal function which provides the interpretable non-linearity (figure \ref{fig:main}B). Overall, this positions our model apart from previous pseudotime approaches that all emphasise linear and nonlinear \emph{unsupervised} learning of the trajectories (figure \ref{fig:main}E).

%In addition, we model dropouts and a strict empirically motivated mean-variance relationship (see Supplementary Information) which is required to provide constraints on the latent variable model since nothing on the right hand side of Equation \ref{eq:master_eq} is actually measured or observed.


\subsubsection{Mean-variance relation}
%
% \noindent\fbox{%
% \parbox{\textwidth}{%
% Reproducible analysis for exploratory plots can be found in \texttt{mean-variance.Rmd} and for parameter estimation in \texttt{generating-synthetic-data.Rmd}.
% }%
% }
% \;

RNA-seq counts are commonly assumed to follow the negative binomial distribution 
\cite{Anders2010,Love2014-lx,Robinson2010-rj}
which for gene $g$ in sample $n$ implies a mean-variance relationship of the form

\begin{equation} \label{eq:overd}
\sigma_{ng}^2 = \mu_{ng}(1 + \phi_g \mu_{ng})
\end{equation}

where $\phi_g$ is a gene-specific dispersion factor. Such strong parametric forms are typically required since for low sample sizes the estimates of the variance can be very unstable. % [reference?].

However, typically in single-cell data there are enough measurements (ie cells) to allow robust estimation of both the mean and variance for each gene\cite{Finak2015-iy}. The exception to this is in pseudotime analyses, where we are assuming each cell represents a unique time point, and therefore the mean $\mu_{cg}$ and variance $\sigma^2_{cg}$ are effectively measured only once. Consequently we must consider a strong mean-variance relationship since assuming a constant variance per gene is akin to under-fitting while it would be impossible to fit a variance for each cell and each gene (since there is only one measurement).

As a solution to this we examine the mean-variance relationship for the genes across all cells and assume the same relationship approximately holds for cells as they progress along pseudotime trajectories. The relationship in \ref{eq:overd} applies to the original untransformed data (e.g. TPM or scaled counts) while we wish to model the $\log_2(\text{TPM} + 1)$ transformed relationship directly. Therefore we must examine the mean-variance relationship for the $\log_2$ data\footnote{Note for $x \gg 0$ $\log_2(x+1) \approx \log_2(x)$.}, since in general the mean-variance relationship of the log-transormed data isn't the same as (the log of) the mean-variance relationship on the untransformed data.

\begin{figure}
\centering
 \includegraphics[width=\textwidth]{gfx/ch3/supp_paper_variance_2}
 \caption{The mean-variance relationship in $\log_2(\text{TPM} + 1)$ single-cell data for the three datasets studied in the main text. Red denotes the marker genes identified in the main text while blue corresponds to three house-keeping genes (\emph{LDHA}, \emph{NONO}, \emph{PGK1}) or their mouse equivalents. Both the Trapnell and Zhou datasets show consistent evidence that pseudotemporal marker genes exist on the `leading edge' of the data with medium mean expression but high variance. This suggests a linear relationship between mean and variance in $\log_2$ space. In contrast, the housekeeping genes all sit in the `tails' with high mean expression but very low variance.} \label{fig:variance}
 \end{figure}

 The mean-variance relationship of $\log$ count data has been examined both in Limma Voom \cite{Law2014-tu}
 and sleuth \cite{Pimentel2016-xz} that conclude there is no closed form mean-variance relationship for logged data. Their solution is to fit a LOESS curve between the empirical mean and variance (or standard deviation) of each gene, and use this to shrink variance estimates towards a common value given the mean expression. However, if possible we would like a fully Bayesian model that fits all parameters simultaneously, rather than pre-fitting variance estimates with LOESS. We therefore consider a parametric approximation.

We examined the mean-variance relationship of the logged data for the three datasets in the main text, as seen in figure \ref{fig:variance}. The pseudotemporal marker genes identified in the original text are shown in red. For both the Trapnell and Zhou datasets these lie on the `leading edge' of the relationship, in areas of moderate mean expression but high variance. In comparison, the housekeeping genes (shown in blue) lie in the tails in regions of high mean expression but low overall variance. This makes intuitive sense, as we expect the marker genes to turn on across the trajectory, giving them a mean of around half the maximal value but maximum variance. In contrast, we expect housekeeping genes to have maximal expression across the trajectory but very low variance in keeping with the constancy of their expression. We therefore assumed that any genes we wish to model as pseudotemporal marker genes follow the same linear mean-variance relationship. 

% Oddly we found such a relationship not to hold in the Shin et al. dataset.


\subsubsection{Single-cell dropout}

% \noindent\fbox{%
% \parbox{\textwidth}{%
% Reproducible analysis for parameter estimation can be found in \texttt{generating-synthetic-data.Rmd}.
% }%
% }
% \;


As previously discussed in section \ref{sec:intro_dropout} single-cell RNA-seq data is known to exhibit substantial dropout, whereby a failure to reverse transcribe lowly expressed transcripts results in zeros in the count matrix. Note that this is distinct from \emph{true zeros} where no mRNA transcripts are truly present, due to either bursty or non-expression.

To account for this we assume that the probability of a dropout is logistic on the latent gene expression mean in a similar approach to Kharchenko et al 2014. The difference to previous approaches is that (1) we are working in $\log_2(\text{TPM} + 1)$ space and (2) we assume a unique mean $\mu_{cg}$ for every gene in every cell, giving a per-gene per-cell dropout probability $p(\mu_{cg})$ of

\begin{equation}
\log\left[ \frac{p(\mu_{cg})}{1 - p(\mu_{cg})} \right] = \beta_0 + \beta_1 \mu_{cg}.
\end{equation}

The dropout parameters $\beta_0$ and $\beta_1$ are constant across all cells as modelling gene specific dropout parameters\footnote{Which is presumably closer to the true model due to gene-specific biases such as GC content.} would require extreme shrinkage across genes in order to make the parameters identifiable.

% To generate synthetic data we fit a logistic regression curve to the probability of dropout against mean $log_2$ expression for the Trapnell et al dataset (figure \ref{fig:dropout}). This gave coefficients of $\beta_0 = 1.763$ and $\beta_1 = -1.156$.

\begin{figure}
\centering
 \includegraphics[width=0.4\textwidth]{gfx/ch3/drop_plot}
 \caption{The probability of a dropout against the mean $\log_2$ expression in the Trapnell et al. dataset. The red solid line shows the logistic regression fit.} \label{fig:dropout}
 \end{figure}

\subsection{Statistical model specification}

%  \subsubsection*{Model Specification}

We now detail the overall specification of our nonlinear factor analysis pseudotime model. As usual we index $N$ cells by $n \in 1, \ldots, N$ and $G$ genes by $g \in 1, \ldots, G$ and let $y_{ng} = [\bY]_{ng}$ denote the (log-transformed and suitably normalised) observed cell-by-gene expression matrix. In order to make the strength parameters comparable between genes we normalise the gene expression so the approximate half-peak expression is 1 through the transformation

\begin{equation}
\mby_g \rightarrow \mby_g / s_g.
\end{equation}

Here we define $s_g$ as a gene-specific size factor that approximates the half-peak expression of gene $g$:

\begin{equation}
s_g = \frac{1}{|\mathcal{Y}_g^*|} \sum_{y_{ng}^* \in \mathcal{Y}_g^*} y_{ng}^*
\end{equation}

where

\begin{equation}
\mathcal{Y}_g^* = \{ y_{ng} : y_{ng} > 0 \}.
\end{equation}

is the set of non-zero measurements for $g$

Our statistical model can be specified as a Bayesian hierarchical model where the likelihood is given by a bimodal distribution formed from a mixture of zero-component (dropout) and an non-zero expressing cell population. Let $\mbt$ be an $N$-length pseudotime vector (one for each cell), then:
\begin{align}
		y_{ng} & \sim \begin{cases}
		\theta_{ng} + (1 - \theta_{ng}) \mathrm{Student}(\mu_{ng}, \sigma^2_{ng}) & \mbox{if } y_{ng} = 0 \\
		 (1 - \theta_{ng}) \mathrm{Student}(\mu_{ng}, \sigma^2_{ng}) & \mbox{if } y_{ng} > 0
		\end{cases}, \\
		\theta_{ng} & \sim \mathrm{Bernoulli}(\mathrm{logit}^{-1} (\beta_0 + \beta_1 \mu_{ng})), \\
		\mbbeta & \sim \norm(0, 0.1) .
\end{align}

The relationship between dropout rate and expression level is expressed as a logistic regression model \cite{Kharchenko2014}. Furthermore, we impose a mean-variance relationship of the form:
\begin{equation}
\sigma^2_{ng}  = (1 + \phi) \mu_{ng} + \epsilon
\end{equation}

with the hierarchical prior structure

\begin{align}
	t_n &\sim \norm(0.5, 1), \\
	\mu_{ng} & =  \mO_g f(t_n, k_g, \tO), \\
	\phi & \sim \mathrm{Gamma}(\alpha_\phi, \beta_\phi)
\end{align}
where $\phi$ is a dispersion parameter. 
%This model is motivated by empirical observations of marker gene behaviour.

We define the sigmoid function as
\begin{align}
	f(t_c;k_g,t^{(0)}_g) &= \frac{2}{1 + \exp\left(-k_g(t_c - t^{(0)}_g)\right)},
\end{align}
where $k_g$ and $\tO_g$ denote the activation strength and activation time parameters for each gene and $\mO_g$ the average peak expression
with default priors
\begin{align}
		\mO_g & \sim \mathrm{Gamma}(\delta / 2, 1 / 2), \\
		k_g & \sim \norm(\mu^{(k)}_g, 1 / \tau^{(k)}_g), \\
		\tO_g & \sim \norm(\mu^{(t)}_g, 1 / \tau^{(t)}_g),
\end{align}
If available, user-supplied prior beliefs can be encoded in these priors by specifying the parameters $\mu^{(k)}_g, \tau^{(k)}_g, \mu^{(t)}_g, \tau^{(t)}_g$. Otherwise, inference can be performed using uninformative hyperpriors on these parameters. Specifying $\mu^{(k)}_g$ encodes a prior belief in the strength and direction of the activation of gene $g$ along the trajectory with $\tau^{(k)}_g$ (inversely-) representing the strength of this belief. Similarly, specifying $\mu^{(t)}_g$ encodes a prior belief of where in the trajectory gene $g$ exhibits behaviour (either turning on or off) with $\tau^{(t)}_g$ encoding the strength of this belief.

\subsection{Inference}

We performed posterior inference using Markov Chain Monte Carlo (MCMC) stochastic simulation algorithms, specifically the No U-Turn Hamiltonian Monte Carlo approach \cite{homan2014no} implemented in the STAN probabilistic programming language \cite{carpenter2015stan} (see section \ref{sec:stan}). The parameter $\epsilon = 0.01$ is used to avoid numerical issues in MCMC computation. For larger marker gene panels, such as in the cell cycle analysis section, we used the automatic differentiation variational inference (ADVI) implemented in STAN to perform approximate Bayesian inference.


\subsubsection{Convergence diagnostics}

\begin{figure}
\centering
 \includegraphics[width=0.9\textwidth]{gfx/ch3//mcmc}
 \caption{\textbf{MCMC convergence diagnostics}. \textbf{A} For a single chain \sname contains a convenience function to plot both the trace of the log-likelihood and the log-likelihood autocorrelation. \textbf{B-E} MCMC diagnostic statistics such as posterior log-likelihood trace, effective sample size, Gelman-Rubin statistic and ratio of monte carlo standard error to posterior standard deviation show good mixing across chains.} \label{fig:mcmc}
 \end{figure}

Since \sname relies on MCMC simulation for inference it is cruicial to assess model convergence before drawing statistical conclusions. The \sname \texttt{R} package contains a convenience function \texttt{plot(ouija\_fit, what = "diagnostic")} to plot both the posterior log-likelihood and posterior log-likelihood autocorrelation, an example of which is in figure \ref{fig:mcmc}A. In general we found it practical to run MCMC inference on only a single chain for most models. Therefore, to check both good mixing of the chains and that the models aren't getting stuck in local maxima, we tested the models for both the Trapnell and Shin datasets across multiple chains. We then examined several diagnostics such as the log-posterior mixing, effective sample size, Gelman-Rubin statistic and ratio of monte carlo standard error to posterior standard deviation (figures \ref{fig:mcmc}B-E), all of which fell inside the acceptable range. 

%Full analysis can be found in the Rmarkdown notebooks online at \url{www.github.com/kieranrcampbell/ouija-paper}.

\subsection{Generating synthetic data}

Our method for generating synthetic pseudotemporally regulated single-cell RNA-seq data representing $\log_2(\text{TPM} + 1)$ given the above considerations can be found in algorithm \ref{alg:synth}.

To quantify this for generating synthetic data we fit a simple linear model with a forced zero intercept to the pseudotemporal marker genes in the Trapnell et al. dataset that gave a gradient of $3.502$. Therefore, in all our synthetic data we model $\sigma^2 = 3.5 \mu$.



\begin{algorithm}                      % enter the algorithm environment
\caption{Generate pseudotemporally regulated scRNA-seq data}          % give the algorithm a caption
\label{alg:synth}                           % and a label for \ref{} commands later in the document
\begin{algorithmic}[1]                    % enter the algorithmic environment
\State \textbf{Data:} $G$ genes, $C$ cells, $\phi = 3.5$, $\beta_0 = 1.76$, $\beta_1 = -1.16$
\State \textbf{Result:} A $C \times G$ matrix of gene expression $X$, where $[X]_{cg} = x_{cg}$
\For{$g \in 1 \ldots G$}
	\State Draw $k_g \sim \text{Unif(5, 10)}$
	\State Draw $\mu_g^{(0)} \sim \text{Unif}(2, 5)$
	\State Draw $t_g^{(0)} \sim \text{Unif}(0,1)$
	\State Set $k_g \gets -k_g$ with probability $\frac{1}{2}$
	\For{$c \in 1 \ldots C$}
		\State Draw $t_c \sim \text{Unif}(0,1)$
		\State $\mu_{cg} \gets \mu_g^{(0)} f(t_c, k_g, t_g^{(0)})$
		\State $\sigma^2_{cg} \gets \phi \mu_{cg}$
		\State Draw $x_{cg} \sim \norm(\mu_{cg}, \sigma_{cg}^2)$
		\If{$x_{cg} < 0$}
			\State $x_{cg} \gets 0$
		\EndIf
		\State $\theta_{cg} \gets \text{logit}^{-1}(\beta_0 + \beta_1 \mu_{cg})$
		\State Set $x_{cg} \gets 0$ with probability $\theta_{cg}$
	\EndFor
\EndFor
\end{algorithmic}
\end{algorithm}



%\section{Analyses performed}
%
%All scripts for figures and analyses available at \url{www.github.com/kieranrcampbell/ouija-paper}.
%
%\subsection{Equivalence of pseudotimes to whole transcriptome sets}
%
%\paragraph{Trapnell et al. 2014}
%
%The dataset was downloaded (11/5/2016) as part of the \texttt{Monocle 2} \texttt{R} package from github via \texttt{devtools::install\_github("cole-trapnell-lab/monocle-release@monocle2")}. The dataset, named \texttt{HSMM}, contains slots for the pseudotimes inferred in \cite{Trapnell2014} in the \texttt{Pseudotime} slot with the cell type inferred in the \texttt{State} slot. Cell types corresponding to ``3'' were identified in the original publication as contaminating interstitial mesenchymal cells and so were removed from the analysis. The values in \texttt{exprs(HSMM)} represent FPKM gene expression data, so for inference with \sname these were $\log_2$-transformed with a pseudocount of 1 added. \sname was run with the default parameters apart from the mean vector $\bm{\mu}^{(k)} = (-10, -10, 10, 10, 10)$ representing prior information. MCMC convergence was assessed by examining the trace and autocorrelation of the log-likelihood.
%
%\paragraph{Shin et al. 2015}
%
%The dataset was downloaded (11/5/2016) from \url{http://www.cell.com/cms/attachment/2038326541/2052521610/mmc7.xlsx}. The original Waterfall pseudotimes were taken from the \texttt{Pseudotime} row, with the remainder of the data representing TPM expression values. Once more the TPM values were $\log_2$ transformed with a pseudocount of 1, and \sname run with the default parameters apart from the mean vector $\bm{\mu}^{(k)} = (5, 5, 5, -5, -5, -5)$ representing prior information. MCMC convergence was assessed by examining the trace and autocorrelation of the log-likelihood.

\subsection{Benchmarking against alternative pseudotime algorithms}

\paragraph{Generating synthetic data}

Data was generated as per algorithm \ref{alg:synth} for 100 cells and 6, 9, 12 and 15 genes, with 40 replicates each.

\paragraph{\sname configuration}

As discussed in the main text, \sname was benchmarked with three different prior configurations. Apart from the parameters mentioned, all others correspond to the \sname defaults. Let $k_g^*$ correspond to the true $k_g$ used in synthetic data generation. Then the prior structures were encoded as follows:

\begin{enumerate}
\item \emph{random} - $\mu^{(k)}_g \gets 8$ with probability $\frac{1}{2}$ and $\mu^{(k)}_g \gets -8$ with probability $\frac{1}{2}$. $\mu^{(t)}_g \gets 0.5$ (default).
\item \emph{directional prior} - $\mu^{(k)}_g \gets 8 \times \text{sign}(k_g^*) $.  $\mu^{(t)}_g \gets 0.5$ (default).
\item \emph{true prior} - $\mu^{(k)}_g \gets k_g^*$, $\mu^{(t)}_g$ set to the true value of $t_g^{(0)}$.
\end{enumerate}


\paragraph{Monocle 2 configuration}

Monocle 2 was run with default parameters with the exception of \texttt{reduceDimension(cds, norm\_method = "none")} to account for the fact that the data already represents $\log_2(\text{TPM} + 1)$ normalised values.

\paragraph{DPT configuration} The MATLAB version of DPT was run with \texttt{nsig=20} and no branch detection.

\paragraph{PCA} The first principal component was calculated in \texttt{R} with \texttt{prcomp(X)\$x[,1]}.

\section{Results}


{\color{red}


\subsection{Systematic comparison of marker-based and whole-transcriptome based pseudotime estimation}

We investigated the use of limited marker gene panels for pseudotime inference using three single cell expression datasets: (i) adult hippocampal quiescent neural stem cells (Shin) \cite{Shin2015}, (ii) differentiating myoblasts (Trapnell) \cite{Trapnell2014} and (iii) embryonic precursor cells into haematopoietic stem cells \cite{zhou2016tracing}. We began by computing pseudotimes using a selection of pseudotime algorithms for each dataset using the small set of 5-6 marker genes identified in the respective studies. We compared marker-based pseudotimes estimated using Ouija, a reference pseudotime based on the first principal component (PC1), TSCAN  \cite{ji2016tscan}, Monocle 2 \cite{qiu2017reversed}, and DPT \cite{haghverdi2016diffusion}. Monocle and TSCAN were run with default parameter settings. For DPT, white noise with standard deviation of $10^{-6}$ of each gene’s standard deviation was added to the expression data to avoid errors concerning cells with identical expression.

%\floatsetup[figure]{style=plain,subcapbesideposition=top}

%\begin{figure}[!t]
%	\sidesubfloat[]{\includegraphics[width=0.8\textwidth]{gfx/ch3/trapnell.png} \label{fig:correlation-a}}
%
%	\sidesubfloat[]{
%		\includegraphics[width=0.8\textwidth]{gfx/ch3/trapnell-fit.png}
%		\label{fig:correlation-b}}
%
%	\caption{\textbf{Comparison of marker gene-based pseudotime estimates across five algorithms.} (A) Pseudotimes reported by the five algorithms do not exhibit strong correlation. (B) Expression level fits reported by each algorithm qualitatively reflect their intrinsic modelling assumptions. Blue line shows a smoothing curve fitted to the pseudotemporally ordered expression values.}
%	\label{fig:correlation}
%\end{figure}

Fig. \ref{fig:correlation}A shows the correlation between pseudotimes obtained for the Trapnell data set whilst equivalent results for the Shin and Zhou data sets can be found in Supp. Fig. 1A and Supp. Fig. 2A respectively. Our analysis shows that pseudotime estimates obtained from marker gene panels differed between pseudotime algorithms with many showing only weak correlation between inferred pseudotimes. However, qualitatively, all algorithms produced plausible gene expression variation over the inferred pseudotimes (Fig. \ref{fig:correlation}B, Supp. Fig. 1B and Supp. Fig. 2B).


\begin{figure}%[!t]
	\centering
	\includegraphics[width=\textwidth]{gfx/ch3/algorithms.png}
	\caption{\textbf{Comparison of pseudotime methods across three scRNA-seq data sets.} We compared pseudotime estimates from three studies \cite{Shin2015,Trapnell2014,zhou2016tracing} using 5 different approaches. In (A) we compared the correlation between marker-based pseudotimes with those obtained from random gene-sets of varying (including the marker gene set), (B) we show the correlation with whole transcriptome (global) pseudotime obtained from all genes }
	\label{fig:comparisons}
\end{figure}

We investigated further by examining the relationship between pseudotimes obtained from marker gene panels, whole transcriptomes and gene set sizes falling between these two extrema. We computed a transcriptome-wide pseudotime where genes retained had a variance in log-expression (TPM/FPKM) greater than 1 and were expressed in at least one cell at log-expression greater than $1$. Furthermore, for each dataset and algorithm combination, we generated additional gene sets that consisted of the marker genes and $N = 1, 5, 10, 20, 50, 100, 500, 1000$ randomly chosen genes from the full transcriptome-wide gene set. In doing so we were able to investigate the continuum between marker-based and whole-transcriptome pseudotime estimation. This was performed for 50 replicates for each $N$. We then compared the results on three metrics: (Pearson) correlation to the marker pseudotimes and correlation to the global pseudotimes.

Our major observation is the striking variability in pseudotime estimates across data sets, gene sets and replicates for all algorithms
(Fig. \ref{fig:comparisons}). As the number of additional genes included in each data set increases, the correlation of the estimated pseudotimes to the reference marker-based pseudotimes typically decreases (Fig. \ref{fig:comparisons}A) whilst the correlation with global pseudotime increases respectively (Fig. \ref{fig:comparisons}B). Note that Ouija is not included in the latter analysis since it is not intended to be applied to whole transcriptome analyses. This behaviour is expected since the forms of variation encoded in the larger gene set are likely to differ from that encoded within the marker genes. However, it is evident that depending on the dataset and genes chosen there is large variability in the consistency of pseudotime algorithm fitting. In other words, there is no one global solution that consistently fits the same pseudotime that will recapitulate the desired behaviour of marker genes.

\begin{figure}%[!t]
	\centering
	\includegraphics[width=\textwidth]{gfx/ch3/de_results.png}
	\caption{\textbf{Comparison of differential expression analyses.} We compared differential expression across pseudotime analyses. In (A)  }
	\label{fig:de_results}
\end{figure}

We finally compared the consistency of differential expression across pseudotime using a recently published method for detecting which genes vary across pseudotime. All p-values were corrected using Benjamini-Hochberg and an FDR threshold of 5\% used to define significance. For each algorithm and number of additional genes we computed the proportion of genes deemed significant found in (a) the differential expression test using the transcriptome-wide pseudotime and (b) the differential expression test using the marker-only pseudotime.

It can be see that as the number of markers included increases the rate of overlap to the transcriptome-wide DE calls also generally increases (Fig. \ref{fig:comparisons}D). Notably, the rate is very low for DPT when the number of included markers is small, which is consistent with the low correlation to marker pseudotimes exhibited in figure xA. In contrast, as the number of included marker genes increases, the rate of overlap to the marker DE calls decreases (Figure xE) as expected. }

\section{Incorporating prior information improves pseudotime inference}

\begin{figure}
	\centering	
	\includegraphics[width=\textwidth]{gfx/Ch3/traj_comparison.png} 
	\caption{{\bf Comparison with alternative pseudotime algorithms using synthetic data.} 
	\textbf{A} Example plots of synthetic gene expression across pseudotime for four genes. The algorithm for generating synthetic pseudotemporally regulated gene expression data may be found in supplementary methods.
	\textbf{B} We compared Ouija against three alternative pseudotemporal methods (principal component 1, Monocle 2 and DPT). We generated synthetic data as described in supplementary methods for gene expression matrices of different sizes ($ G = 6, 9, 12, 15$) with forty replicates in each and attempted to re-infer the pseudotimes. We used both Pearson correlation and Kendall's tau as measures of accuracy compared to the true values. Results show that PC1, Monocle 2, DPT and Ouija with non-informative priors perform comparably. Ouija with priors given by directional and true priors are significantly more accurate in all cases.}
	\label{fig:benchmarking}
\end{figure}

To compare the performance of Ouija to other pseudotemporal ordering algorithms on small gene sets we created synthetic gene expression data with known pseudotimes then re-inferred the pseudotimes using the algorithms. Briefly, pseudotimes were generated on a uniform distribution on $[0, 1)$ and $k_g$, $\mu^{(0)}_g$ and $t^{(0)}_g$ generated for each gene. Gene expression representing $\log_2(\text{TPM + 1})$ was then generated using a Gaussian likelihood censored from below at zero. Care was taken to ensure the mean-variance and dropout relations were comparable to real single-cell RNA-seq data. Full details can be found in supplementary methods.

To benchmark the algorithms we generated data for $C = 100$ cells for $G \in \{6, 9, 12, 15\}$ genes, with forty replicates of each dataset for varying values of $k$, $t_0$ and $\mu_0$. We benchmarked Ouija against principal component 1 (PC1), Diffusion Pseudotime (DPT) \cite{haghverdi2016diffusion} and Monocle 2 \cite{Trapnell2014}. Ouija allows us to specify priors on 	$k$ and $t_0$ via $k_g \sim \norm(\mu^{(k)}_g, 1 / \tau^{(k)}_g)$ and $\tO_g \sim \norm(\mu^{(t)}_g, 1 / \tau^{(t)}_g)$. We benchmarked Ouija in three situations. Firstly, with \emph{random priors} on $k$ where $\mu^{(k)}_g$ is set to arbitrary values and $\mu^{(t)}_g$ is set to 0.5 (essentially non-informative) for all $g$. Secondly, \emph{directional priors} where $\mu^{(k)}_g$ is set to a constant multiplied by the sign of the true $k_g$ and $\mu^{(t)}_g = 0.5$. This situation is most likely to correspond to the real-life use case where the researcher knows the direction of change of the genes along the trajectory along with a guess of the magnitude of that change. Finally, we considered priors with \emph{true values} where $\mu^{(k)}_g$ and $\mu^{(t)}_g$ are set to the true values used in the synthetic data generation. In general it is highly unrealistic for researchers to know the true parameter values but we include this case for comparison to the more realistic scenario of directional priors. In all situations the precision parameters are set to 1.

The results can be seen in Figure \ref{fig:benchmarking}. Performance was assessed using Pearson correlation to the true pseudotimes as well as Kendall's tau, a measure of concordance of the ordering of two sets. For both metrics three trends are obvious. Firstly, as the number of genes in the marker panel increases, the accuracy of the pseudotemporal inference also increases. Secondly, PCA, DPT, Monocle 2 and Ouija with random priors are all largely comparable in their results, with Ouija with both directional priors and true values performing significantly better. Finally, the performance of Ouija with directional priors and the true values are largely comparable, implying exact knowledge of the biological parameters isn't necessary for the inclusion of prior information to be useful.




\subsection{Ouija clusters cell types based on pseudotime continuity}

We further investigated the single cell expression data from a study tracking the differentiation of embryonic precursor cells into haematopoietic stem cells (HSCs) \cite{zhou2016tracing}. The cells begin as haemogenic endothelial cells (ECs) before successively transforming  into pre-HSC and finally HSC cells. The authors identified six marker genes that would be down-regulated along the differentiation trajectory, with early down-regulation of \emph{Nrp2} and \emph{Nr2f2} as the cells transform from ECs into pre-HSCs, and late down-regulation of \emph{Nrp1}, \emph{Hey1}, \emph{Efnb2} and \emph{Ephb4} as the cells emerge from pre-HSCs to become HSCs. The study investigated a number of distinct cell types at different stages of differentiation: EC cells, T1 cells ($CDK45^-$ pre-HSCs), T2 cells ($CDK45^+$ pre-HSCs) and HSC cells at the E12 and E14 developmental stages.

We conducted a pseudotime analysis using Ouija on the 105 cells featured in the original experiment to investigate whether the inferred pseudotime progression could recapitulate the known cell types in the study and their known relationships in an unsupervised analysis using just these six marker genes alone. As Ouija uses a probabilistic model and inference we were able to obtain a posterior ordering matrix (Fig. \ref{fig:hsc}A) that describes the probability of the ordering of any two cells. When cells are ordered by the expected pseudotime, this posterior matrix contained three metastable groups of cells corresponding to endothelial, pre-HSCs and HSCs respectively (Fig. \ref{fig:hsc}B). Misclassifications within cell types (T1/T2 and E12/E14 cells) could be explained by examining a principal components analysis of the global expression profiles (Fig. \ref{fig:hsc}C) which suggests that these cell types are not completely distinct in terms of expression.

When examining the inferred pseudotime progression of each marker gene (Fig. \ref{fig:hsc}D), these three metastable states corresponded to the activation of all genes at the beginning of pseudotime time, the complete inactivation of all the marker genes at the end of the pseudotime and a intervening transitory period as each marker gene turns off. Each metastable state clearly associates with a particular cell type (Fig. \ref{fig:hsc}E). With, as expected \emph{Nrp2} and \emph{Nr2f2} exhibiting early down-regulation and \emph{Nrp1}, \emph{Hey1}, \emph{Efnb2} and \emph{Ephb4} all exhibiting late down-regulation. Using this HSC formation system as a proof-of-principle it is evident that, if a small number of switch-like marker genes are known, it is possible to recover signatures of temporal progression using Ouija and that these trajectories are compatible with real biology.

\begin{figure}
	\centering
	\includegraphics[width=\textwidth]{gfx/ch3/hsc}
	\caption{{\bf Pseudotime ordering and cell type identification of haematopoeietic stem cell differentiation}
(A) Consistency matrix of pseudotime ordering. Entry in the $i^{th}$ row and $j^{th}$ column is the proportion of times cell $i$ was ordered before cell $j$ in the MCMC posterior traces. Gaussian mixture modelling on the first principal component of the matrix identified three clusters that are evident in the heatmap.
(B) Confusion matrix for cell types identified in original study (rows) and Ouija inferred (columns). Ouija inferred cluster 1 largely corresponds to EC cells, cluster 2 corresponds to pre-HSC cells while cluster 3 corresponds to HSC cells.
(C) PCA plot similar to the original publication \cite{zhou2016tracing} suggests supports the existence of three distinct cell types in the data.
(D) HSC gene expression as a function of pseudotime ordering for six marker genes. Background colour denotes the maximum likelihood estimate for the Ouija inferred cell type in that region of pseudotime.
(E) HSC gene expression as a function of pseudotime ordering for six marker genes with cells coloured by known cell type.
}
	\label{fig:hsc}
\end{figure}


\subsection{Cell cycle prediction as a pseudotime estimation problem}

We wanted to consider a study composed of a large panel of marker genes and identified a single-cell RNA-seq study \cite{kowalczyk2015single} that examined variation between individual hematopoietic stem and progenitor cells from two mouse strains (C57BL/6 and DBA/2) as they age. Principal component analysis for each cell type and age showed a striking association of the top principal components with cell cycle-related genes (Fig. \ref{fig:regev}A), indicating that transcriptional heterogeneity was dominated by cell cycle status. They scored each cell for its likely cell cycle phase using signatures based on functional annotations \cite{reference2009gene} and profiles from synchronized HeLa cells\cite{whitfield2002identification} for the G1/S, S, G2, and G2/M phases.

We investigated if Ouija could be used to identify cell cycle phase, treating the inferential problem as a continuous pseudotime process. We applied Ouija to 1,008 C57Bl/6 HSCs using 374 GO cell cycle genes that satisfied gene selection criteria used in the original study. The estimated pseudotime progression given by Ouija recapitulates the  trajectory observed in principal component space. The estimated pseudotime distribution correlates well with the cell cycle phase categorisation  given in the original study (Fig. \ref{fig:regev}C). Furthermore, we identified 88 genes with large (negative) activation strengths indicating strong switch (off) behaviour (Fig. \ref{fig:regev}D) ordering the cells according to pseudotime and then ordering by activation time shows a cascade of expression inactivation across these 88 genes over cell cycle progression with the quiescent ($G_0$) indicated by complete inactivation of all 88 genes (Fig. \ref{fig:regev}E,F). The explicit parametric model assumed by Ouija makes this gene selection and ordering process simple and \emph{quantitative} compared to a non-parametric approach that would require some retrospective analysis or visual inspection.

This investigation indicates that although Ouija makes strong assumptions about gene-specific expression behaviour, its utility is not limited to small marker gene panels that strictly obey its switch-like behavioural assumptions.

\begin{figure}
	\includegraphics[width=\textwidth]{gfx/ch3/regev.png}
	\caption{{\bf Cell cycle phase prediction.} Principal component representation of hematopoietic stem cells coloured according to (A) the original cell cycle progression score \cite{kowalczyk2015single} and (B) Ouija - cell cycle classes indicated are based on original study classifications. (C) Distribution of Ouija inferred pseudotime versus the original cell cycle classifications. (D) Estimated activation strengths for the 374 cell cycle gene panels. (E) Gene expression profile for 88 switch-like genes with cells ordered by pseudotime and (F) genes ordered by activation time.}
	\label{fig:regev}
\end{figure}


\subsection{Robustness to transient gene behaviour}

Finally, a potential limitation of our switch behaviour model is that it assumes that all selected marker genes follow a strict monotonically increasing (or decreasing) behaviour and there exists a smooth, non-transient transition from an initial cell expression state to a final resting state at the end of the pseudotemporal period. In selecting a marker gene panel the investigator may not always be fully certain of the quantitative behaviour of all genes and some may indeed exhibit a transient rather than switch behaviour. In order to test of the impact of such genes on Ouija we performed a simulation study to assess the effect of genes exhibiting transient behaviour on pseudotime estimation. We did this by simulating panels of single cell expression values that mimic the zero-inflated properties observed in real data for a variety of genes containing mixed numbers of switch-like and transient gene behaviours (Fig. \ref{fig:switch-transient-simulation}A). Specifically we considered two scenarios, the first in which the numbers of switch-like and transient genes are equal and the second in which three-quarters of the gene panels were switch-like genes and the reminder transient (Figure \ref{fig:switch-transient-simulation}B). Our simulation study showed that if the switch-like genes remained the dominant class of genes in the simulated marker gene panels, it remained possible to accurately infer pseudotime in the presence of transient genes (Fig. \ref{fig:switch-transient-simulation}C). Furthermore, as the size of the panel increases, the absolute number of switch genes was a greater determinant of pseudotime estimation accuracy than the proportion of the marker gene panel that was truly switch-like.

\begin{figure}
	\includegraphics[width=\textwidth]{gfx/ch3/switch-transient-simulation.png}
	\caption{{\bf Impact of marker genes exhibiting transient rather than switch-like gene behaviours.} (A) Examples of simulated switch and transient expression genes. (B) Example 24-gene panel with 12 switch and 12 transient genes. (C) Correlation between true and estimated pseudotime as a function of the gene panel size and proportion of switch genes for ten replicates.}
	\label{fig:switch-transient-simulation}
\end{figure}

\section{Discussion}

We have developed a novel approach for pseudotime estimation based on modelling switching expression behaviour over time for marker genes. Our strategy provides an orthogonal and complimentary approach to unsupervised, whole transcriptome methods that do not explicitly model any gene-specific behaviours and do not readily permit the inclusion of prior knowledge.

We demonstrate that the selection of a few marker genes allows comparable pseudotime estimates to whole transcriptome methods on real single cell data sets. Furthermore, using a parametric gene behaviour model and full Bayesian inference we are able to recover posterior uncertainty information about key parameters, such as the gene activation time, that allows us to explicitly determine a potential ordering of gene (de)activation events over (pseudo)time. The posterior ordering uncertainty can also be used to identify homogeneous phases of transcriptional activity that might correspond to transient, but discrete, cell states.

Although our focus was on switching expression behaviour, alternative parametric functions could be used to capture other gene behaviours. However, it is critical to recognise that in a latent variable modelling framework such as this, prior information has a strong influence over the final outcome. Therefore any constraints should match \emph{a priori} knowledge of the marker genes under investigation. Otherwise the pseudotime estimation could forcibly impose a pre-specified temporal form on the data. Furthermore, whilst we do not explicitly address branching processes in this work, our framework provides a natural and simple extension to allow for multiple lineages and cell fates using a sparse mixture of factor analyzers in which each lineage is denoted by a separate mixture component and the factors loadings are shrunk to common values to denote shared branches. This will be developed in future work.

In summary, Ouija provides a novel contribution to the increasing plethora of pseudotime estimation methods available for single cell gene expression data.
